frontend/src/pages/Groups/GroupPage/GroupPage.tsx

```diff
@@ -91,7 +99,55 @@ function GroupHub({ group }: GroupHubProps) {
     }
     else {
       alert("Error when attempting to leave group");
+      push("/groups");
+    }
+  }
+
+
+  async function checkMembership() {
+    if (member) {
+      setLoading(false);
+      return
+    }
+    try {
+      const payload = {
+        user: user?.id,
+        group: id
+      };
+      const memberResp = await checkMembershipMutation({ variables: payload });
+      if (memberResp?.errors) {
+        alert(memberResp.errors);
+        push("/groups");
+      }
+      if (memberResp.data.checkMembership.member === true) {
+        setMember(true);
+        setLoading(false);
+        return
+      }
+      // not a member
+      const joinResp = await joinGroupMutation({ variables: payload });
+      if (joinResp?.errors) {
+        alert(memberResp.errors);
+        push("/groups");
+      }
+      if (joinResp.data.joinGroup.success === true) {
+        setMember(true);
+        setLoading(false);
+        return
+      }
+      console.log("Error trying to join group.")
     }
+    catch(error) {
+      alert(error);
+    }
+  }
+
+
+  if (!member) {
+    checkMembership();
+  }
```

Comment:
We might want to refactor the check membership function. Currently, you are checking member on every render, which isn't too bad, but the React way to handle this would be to check once when the component mounts. This is feasible since the user cannot leave a group and stay on the page. The code would look something like this:

```
useEffect(() => {
setLoading(true)
await checkMembershipMutation({ variables: payload });    

// If there is an error  
if (memberResp?.errors) {        
alert(memberResp.errors);        
push("/groups");      
} 

//  Is a member of the group   
if (memberResp.data.checkMembership.member) { 
setLoading(false);        
return      
}

// Is not a member and is joining
const joinResp = await joinGroupMutation({ variables: payload });
      if (joinResp?.errors) {
        alert(memberResp.errors);
        push("/groups");
      }
      if (joinResp.data.joinGroup.success) {
        setLoading(false);
        return
      }

}, [])
```

This lets us take out the `member` variable which isn't entirely necessary and improves performance. Usually anything asynchronous should be within a callback or a `useEffect` hook. Having it in the render body of a component can lead to render stalls.

Reply:
Note that you'll have to wrap the contents of the `useEffect` in another function located inside the `useEffect` closure since the `useEffect` callback itself cannot have an `async` tag applied to it. For more info see [this](https://stackoverflow.com/questions/53332321/react-hook-warnings-for-async-function-in-useeffect-useeffect-function-must-ret)

Reply:
Just updated, let me know if my recent refactor uses useEffect correctly.